src/single_random_draw.rs

```diff
@@ -0,0 +1,450 @@
+//! This library provides efficient algorithms to compose a set of unspent transaction outputs
+//! (UTXOs).
+
+use crate::errors::Error;
+use crate::CHANGE_LOWER;
+use bitcoin::blockdata::transaction;
+use bitcoin::Amount;
+use bitcoin::FeeRate;
+use bitcoin::ScriptBuf;
+use bitcoin::Transaction;
+use bitcoin::TxIn;
+use bitcoin::TxOut;
+use bitcoin::Weight;
+use rand::seq::SliceRandom;
+
+/// Calculates the effective_value of an input.
+///
+/// Returns `Ok(None)` if the effective_value is negative.  If the effective_value is positive, return `Ok(Some(Amount))`.
+///
+/// ## Errors
+///
+/// Returns `Err(Error::Multiplication)` if `FeeRate` * `Weight` overflows.
+///
+fn get_effective_value(
+    utxo: &TxOut,
+    input: &TxIn,
+    fee_rate: FeeRate,
+) -> Result<Option<Amount>, Error> {
+    let weight_predict = transaction::InputWeightPrediction::new(
+        input.script_sig.len(),
+        input.witness.iter().map(|elem| elem.len()),
+    );
+
+    let input_weight = transaction::predict_weight(Some(vec![weight_predict]), None::<Vec<_>>);
+    let input_fee: Option<Amount> = fee_rate.checked_mul_by_weight(input_weight);
+
+    match input_fee {
+        Some(f) => Ok(utxo.value.checked_sub(f)),
+        None => Err(Error::Multiplication(input_weight, fee_rate)),
+    }
+}
+
+/// Calculate the fee for a transaction with two UTXOs and no inputs.
```

Comment:
Should “a transaction with two UTXOs and no inputs” be “a transaction with two outputs and no inputs”?

Reply:
yes, good point.  thanks!

Reply:
resolved